crankshaft/compiler.md

```diff
@@ -0,0 +1,281 @@
+  - monomorphic functions/constructors can be inlined entirely
+  - inlining allows even further optimizations
+
+### Deoptimization
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=36m50s) | [slide](http://v8-io12.appspot.com/index.html#78)
+
+- optimizations are speculative and assumptions are made
+- if assumption is violated
+  - function deoptimized
+  - execution resumes in full compiler code
+  - in short term execution slows down
+  - normal to occur
+  - more info about about function collected
+  - *better* optimization attempted
+  - if assumptions are violated again, deoptimized again and start over 
+- too many deoptimizations cause function to be sent to *deoptimization hell*
+  - considered not optimizable and no optimization is **ever** attempted again
+- certain constructs like `try/catch` are considered not optimizable and functions containing it go straight to
+  *deoptimization hell* due to **bailout** [watch](http://youtu.be/UJPdhx5zTaw?t=35m23s)
+
+None of this can be diagnosed with Chrome Devtools at this point.
+
+### Causes for Deoptimization
+
+#### Modifying Object Shape
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=21m00s)
+
+- added fields (order matters) to object generate id of hidden class
+- adding more fields later on generates new class id which results in code using Point that now gets Point' to be
+  deoptimized
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=21m45s)
+[watch](http://youtu.be/UJPdhx5zTaw?t=12m18s)
+
+```js
+function Point(x, y) {
+  this.x = x;
+  this.y = y;
+}
+
+var p = new Point(1, 2); // => hidden Point class created
+
+// ....
+
+p.z = 3;                 // => another hidden class (Point') created
+```
+
+- `Point` class created, code still deoptimized
+- functions that have `Point` argument are optimized
+- `z` property added which causes `Point'` class to be created
+- functions that get passed `Point'` but were optimized for `Point` get deoptimized
+- later functions get optimized again, this time supporting `Point` and `Point'` as argument
+- [detailed explanation](http://v8-io12.appspot.com/index.html#30)
+
+##### Considerations
+
+- avoid hidden class changes
+- initialize all members in **constructor function** and **in the same order**
+
+## Efficiently Representing Values and Tagging
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=15m35s) | [slide](http://v8-io12.appspot.com/index.html#34)
+
+- v8 passes around 32bit numbers to represent all values
+- bottom bit reserved as tag to signify if value is a SMI (small integer) or a pointer to an object
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=10m05ss) | [slide](http://v8-io12.appspot.com/index.html#35)
+
+```
+| object pointer              | 1 |
+
+or
+
+| 31-bit-signed integer (SMI) | 0 |
+```
+
+- numbers bigger than 31 bits are boxed 
+- stored inside an object referenced via a pointer 
+- adds extra overhead (at a minimum an extra lookup)
+
+### Considerations
+
+- prefer SMIs for numeric values whenever possible
```

Comment:
also, if a non-SMI is used in a calculation the result will be non-SMI even if the result is in bounds (and then even if it's in bounds and it's considered non-SMI it will continue to make results non-SMI)

```js
var n = 2147483647
%IsSmi(n) // true
n++
%IsSmi(n) // false
n--
%IsSmi(n) // false
%IsSmi(n - 1) // false
```

Reply:
That is an interesting edge-case, however I feel that it adds more confusion to include that in the document (there are lots of things I left out to keep these consumable and understandable).
Also in a way that's explained as part of [double unboxing](https://github.com/thlorenz/v8-perf/blob/turbo-updates/data-types.md#double-array-unboxing) at least for Arrays.

Maybe we could add a note that this applies to individual variables as well? Do you have a suggestion to succinctly state that there?

* * *

WRT https://github.com/thlorenz/v8-perf/pull/14#discussion_r181133693 (again github doesn't let me comment).
Yes I'd think it's bit early to talk about BigInt. Also I didn't find any v8 dev publications about that topic.